framework/geometry.sage

```diff
@@ -89,53 +113,56 @@ def lattice_orthogonal_section(D, V):
     - go back to the primal.
     """
 
-    V = project_and_eliminate_dep(D, V)
+    if not assume_full_rank:
+        V = project_and_eliminate_dep(D, V)
     r = V.nrows()
 
     # Project the dual basis orthogonally to v
     PV = projection_matrix(V)
     D = D - D * PV
 
     # Eliminate linear dependencies
-    D = D.LLL()
-    D = D[r:]
+    if output_basis:
+        D = eliminate_linear_dependencies(D, nb_dep=r)
 
     # Go back to the primal
     return D
 
 
-def lattice_project_against(B, V):
+def lattice_project_against(B, V, assume_full_rank=False, test_belonging=True, output_basis=True):
     """
     Compute the projection of the lattice L(B) orthogonally to Span(V). All vectors if V
-    (or at least their projection on Span(B)) must belong to L(B). 
+    (or at least their projection on Span(B)) must belong to L(B).
     Algorithm:
     - project V onto Span(B)
     - project the basis onto orth(V)
     - eliminate linear dependencies (LLL)
     """
     # Project v on Span(B)
-    V = project_and_eliminate_dep(B, V)
+    if not assume_full_rank:
+        V = project_and_eliminate_dep(B, V)
     r = V.nrows()
 
-    # Check that V belogs to L(B)
-    D = dual_basis(B)
-    M = D * V.T
-    if not lcm([x.denominator() for x in M.list()]) == 1:
-        raise ValueError("Not in the lattice")
+    # Check that V belongs to L(B)
+    if test_belonging:
+        D = dual_basis(B)
+        M = D * V.T
+        if not lcm([x.denominator() for x in M.list()]) == 1:
+            raise ValueError("Not in the lattice")
 
     # Project the basis orthogonally to v
     PV = projection_matrix(V)
     B = B - B * PV
 
     # Eliminate linear dependencies
-    B = B.LLL()
-    B = B[r:]
+    if output_basis:
+        B = eliminate_linear_dependencies(B, nb_dep=r)
 
     # Go back to the primal
     return B
 
 
-def lattice_modular_intersection(D, V, k):
+def lattice_modular_intersection(D, V, k, assume_full_rank=False, output_basis=True):
```

Comment:
Those option could be documented a bit (explain that for optimization purpose, output_basis=False only gives a generator set rather than a basis to save some LLL calls)

Reply:
Yes, I will add more documentation.